contracts/StorageManager.sol

```diff
@@ -46,48 +49,63 @@ contract StorageManager {
+    event AgreementFundsPayout(bytes32 indexed agreementReference, uint256 amount, address indexed token);
     event AgreementStopped(bytes32 indexed agreementReference);
 
+    constructor() Ownable() public { }
+
+    /**
+    @notice whitelist a token or remove the token from whitelist
+    @param token the token from whom you want to set the whitelisted
+    @param isWhiteListed whether you want to whitelist the token or put it from the whitelist.
+    */
+    function setWhitelistedTokens(address token, bool isWhiteListed) public onlyOwner {
+        isWhitelistedToken[token] = isWhiteListed;
+    }
+
+
     /**
     >> FOR PROVIDER
     @notice set the totalCapacity and billingPlans of a Offer.
     @dev
     - Use this function when initiating an Offer or when the users wants to change more than one parameter at once.
     - make sure that any period * prices does not cause an overflow, as this can never be accepted (REF_MAX_PRICE) and hence is pointless
+    - only whitelisted tokens are allowed to make a token for
+    - if there are two tokens, and two billingPrice/periods pairs per token, then boundaries[0] == 1. 
+      This makes the first two billingPeriod/prices pairs to apply to the first token ([tokens[0]]) and the second pairs to the second token.
     @param capacity the amount of bytes offered. If already active before and set to 0, existing contracts can't be prolonged / re-started, no new contracts can be started.
-    @param billingPeriods the offered periods. Length must be equal to the lenght of billingPrices.
-    @param billingPrices the prices for the offered periods. Each entry at index corresponds to the same index at periods. When a price is 0, the matching period is not offered.
+    @param billingPeriods the offered periods. Length must be equal to the length of billingPrices. The first index of the multi dem array corresponds with the address in tokens at the same index
+    @param billingPrices the prices for the offered periods. Each entry at index corresponds to the same index at periods. The first index of the multi dem array corresponds with the address in tokens at the same index
+    @param tokens the tokens for which an offer is made. By convention, address(0) is the native currency.
     @param message the Provider may include a message (e.g. his nodeID).  Message should be structured such that the first two bits specify the message type, followed with the message). 0x01 == nodeID
     */
     function setOffer(uint128 capacity,
-        uint64[] memory billingPeriods,
-        uint64[] memory billingPrices,
+        uint64[][] memory billingPeriods,
+        uint64[][] memory billingPrices,
+        address[] memory tokens,
         bytes32[] memory message
     ) public {
         Offer storage offer = offerRegistry[msg.sender];
         setTotalCapacity(capacity);
-        require(billingPeriods.length > 0, "StorageManager: Offer needs some billing plans");
```

Comment:
To verify @Eknir , why is this removed?

Reply:
Ha @julianmrodri , I am talking to myself as I was reviewing the changeset. It was removed but while removing couldn't catch why exactly. I'll review this and report here.

Reply:
I verified. If the billingPeriod for a particular token is not mentioned, the token will just be disregarded and no state changes will be made (line 372).

Reply:
Done